### PR TITLE
N°3685 - Improve external resources usage

### PR DESCRIPTION
**Pull request by Molkobain**
_DISCLAIMER: This PR is a medium for an internal discussion within the R&D team at Combodo, hence it does not contains detailed information about what it does / aims to._

## Optimise included files
- [X] Identify JS / CSS files that are no longer used in iTop Community (deprecated)
  - Move them in `\WebPage::COMPATIBILITY_DEPRECATED_LINKED_SCRIPTS_REL_PATH` / `\WebPage::COMPATIBILITY_DEPRECATED_LINKED_STYLESHEETS_REL_PATH`
  - Add possibility to include them back via a conf. param. `compatibility.include_deprecated_js_files` / `compatibility.include_deprecated_css_files`
  - For the first time we remove deprecated stuff without notice (we usually give a 1-2 versions delay) but
    - It's necessary to get back to better/equivalent performances
    - They can be included back if necessary

- [X] Identify JS / CSS files that should not be included in all pages
  - Move them in `\WebPage::COMPATIBILITY_MOVED_LINKED_SCRIPTS_REL_PATH` / `\WebPage::COMPATIBILITY_MOVED_LINKED_STYLESHEETS_REL_PATH`
  - Add possiblity to include them back via a conf. param. `compatibility.include_moved_js_files` / `compatibility.include_moved_css_files`

- [X] Make a class to ease usage of third party libs (include all their files, snippets, ...)
  - Ease usage in different places of the app (instead of duplicating)
  - Ease maintenance by avoiding to change an include only in one place
  - Example of usage `ExternalResourcesHelper::AddCKEditorToWebPage(\WebPage &$oPage)`
  - [X] Use it for D3/C3.js
  - [X] Use it for CKEditor
  - [X] use it for SimpleGraph

- [x] Identify Combodo's extensions impacted by those files removal

## Optimise main CSS
- [x] Include only necessary parts of the Bulma framework
  - For now we use only base/_all, elements/_content. Other parts necessary?
  - Gained 590kB / 67% of the main.css file (880kB => 290kB)
  - Remaining Bulma parts weight about 20% of the 290kB
- [x] Improve usage of the `.ibo-button` selector to limit duplications due to `@extend` and loops
	
## Summary of improvements
- JS: Around 30 files (re)moved / 1.6 MB less per page (versus 3.0.0-beta3)
- CSS: 1-2 files (re)moved / 700kB less per page (versus 3.0.0-beta3)
- Fonts: Migrate to WOFF / 500kB less per page (versus 3.0.0-beta3)

## Impacted extensions
Extension | What to do | Status
-----------|--------------|--------
Customer survey | Call `WebResourcesHelper::EnableCKEditor()` only in 3.0+ | ⌛ In progress
Template base | `ConsoleSimpleFieldRenderer` should be migrated (not really in the extension actually), need to be tested | ✔ Tested / OK
	
## TODOs
- [x] Create `ExternalResourcesHelper` class and its tests
- [x] Add config. params. to the wiki
- [x] Switch to WOFF(2) fonts for Raleway

### CODE REVIEW

**Comment by Molkobain** [comment]
Only tests are missing now; will try to write them tomorrow. Sounds like quite an improvement compared to 3.0.0-beta3, congrats everybody! 🙌

File: css/backoffice/components/_button-group.scss [hunk 1]
```diff
@@ -45,8 +45,10 @@ $ibo-button-group--elements-separator--border-left: 1px solid transparent !defau
     border-left: $ibo-button-group--elements-separator--border-left;
   }
 
-  /* Note: Selector cannot be simplified in the one above as the "::before" must be after the different classes */
-  .ibo-button + .ibo-button {
+  /* Note: Selector cannot be simplified in the one above as the "::before" must be after the different classes.
+  * A more precise selector could be .ibo-button + .ibo-button, however .ibo-button is extended by multiple element and in a multiple loop it creates heavy selectors.
+  */
+  > * + * {
```
Comment (piRGoif) [inline css/backoffice/components/_button-group.scss:54]: You should document those CSS selectors changes in the PR's description

File: datamodels/2.x/itop-hub-connector/js/hub.js [hunk 2]
```diff
@@ -2,7 +2,7 @@
 $(function()
 {
 	// the widget definition, where "itop" is the namespace,
-	// "fieldsorter" the widget name
+	// "hub_installation" the widget name
```
Comment (piRGoif) [inline datamodels/2.x/itop-hub-connector/js/hub.js:5]: This is not really related to this PR but... good catch :)

File: js/datatable.js [hunk 3]
```diff
@@ -42,7 +42,6 @@ $(function()
 			$('#sfl_'+sListId).fieldsorter({hasKeyColumn: bViewLink, labels: this.options.oLabels, fields: this.options.oColumns, onChange: function() { me._onSpecificSettings(); } });
 			$('#datatable_dlg_'+sListId).find('input[name=page_size]').on('click', function() { me._onSpecificSettings(); });
 			$('#datatable_dlg_'+sListId).find('input[name=save_settings]').on('click', function() { me._updateSaveScope(); });
-			this.element.find('.itop_popup > ul li').popupmenu();
```
Comment (piRGoif) [inline js/datatable.js:45]: I didn't find where this was moved. Was it just removed ?

File: sources/application/WebPage/WebPage.php [hunk 4]
```diff
@@ -851,6 +915,55 @@ public function add_linked_stylesheet($s_linked_stylesheet, $s_condition = "")
...
+			$this->AddCompatibilityFiles(static::ENUM_COMPATIBILITY_FILE_TYPE_CSS, static::ENUM_COMPATIBILITY_MODE_MOVED_FILES);
+		}
+	}
+
+	/**
+	 * Add compatibility files of $sFileType from $sMode (which are declared in {@see static::COMPATIBILITY_<MODE>_LINKED_<TYPE>_REL_PATH}) back to the page
+	 *
+	 * @param string $sFileType {@see static::ENUM_COMPATIBILITY_FILE_TYPE_XXX}
+	 * @param string $sMode {@see static::ENUM_COMPATIBILITY_MODE_XXX}
+	 *
```
Comment (piRGoif) [inline sources/application/WebPage/WebPage.php:954]: Maybe add 2 `@uses` to indicate which contants and methods are called ?

File: sources/application/WebPage/NiceWebPage.php [hunk 5]
```diff
@@ -22,7 +22,48 @@
...
+		'js/clipboard.min.js',
+		'js/clipboardwidget.js',
+		// - SearchForm
+		'js/searchformforeignkeys.js',
+	];
+	/** @inheritDoc */
+	protected const COMPATIBILITY_DEPRECATED_LINKED_SCRIPTS_REL_PATH = [
+		/** @deprecated since 3.0.0 Not used in the backoffice since the introduction of the new tooltip lib. */
+		'js/hovertip.js',
+		/** @deprecated since 3.0.0 N°2737 - Migrate table to DataTables plugin to be iso with the end-users portal, will be removed in 3.x */
```
Comment (piRGoif) [inline sources/application/WebPage/NiceWebPage.php]: Remove the "since"
Check syntax ref : https://docs.phpdoc.org/3.0/guide/references/phpdoc/tags/deprecated.html#deprecated

**Review by piRGoif — DISMISSED** [review]

File: js/datatable.js (hunk 3 above)
Comment (Molkobain) [reply js/datatable.js:45]: The JS lib is still used in the impact analysis for example, but in the dataTables we now use the PopoverMenu block.

File: sources/application/WebPage/WebPage.php (hunk 4 above)
Comment (Molkobain) [reply sources/application/WebPage/WebPage.php:954]: Not sure to understand, you want me to replace the `@see` or to replace the stated constants? I did not put all the declinaisons of the constants on purpose to avoid flooding.

File: sources/application/WebPage/NiceWebPage.php (hunk 5 above)
Comment (Molkobain) [reply sources/application/WebPage/NiceWebPage.php]: Done :)

File: sources/application/WebPage/WebPage.php (hunk 4 above)
Comment (piRGoif) [reply sources/application/WebPage/WebPage.php:954]: Sorry, I should have wrote a suggestion
Here it is !

```suggestion
	 *
	 * @uses COMPATIBILITY_*_LINKED_STYLESHEETS
	 * @uses COMPATIBILITY_*_LINKED_SCRIPTS
	 * @uses add_linked_add_linked_stylesheet
	 * @uses add_linked_add_linked_script
	 *
```

File: sources/application/WebPage/NiceWebPage.php (hunk 5 above)
Comment (piRGoif) [reply sources/application/WebPage/NiceWebPage.php]: 👌

File: sources/application/WebPage/WebPage.php (hunk 4 above)
Comment (Molkobain) [reply sources/application/WebPage/WebPage.php:954]: Done. After our (IRL) discussion I moved the comment to the `@uses add_linked_*` lines as the $sFileType determines the method called but is not used as-is in the method name.

Comment (piRGoif) [reply sources/application/WebPage/WebPage.php:954]: 👌

**Review by steffunky — DISMISSED** [review]
